idaes/core/scaling/autoscaling.py

```diff
@@ -0,0 +1,215 @@
+    ):
+        """
+        Calculate scaling factors for all constraints in a model based on the norm of
+        the Jacobian matrix, accounting for any variable scaling factors.
+
+        Args:
+            blk_or_cons - block or constraint to calculate scaling factors for
+            norm - type of norm to use for scaling. Must be a positive integer.
+            descend_into - if blk_or_cons is a Block, whether to descend into any sub-Blocks
+              (default=True)
+
+        Returns:
+            None
+
+        Raises:
+            TypeError is blk_or_cons is not a Block or Constraint
+            ValueError if norm is not a positive integer
+        """
+        # Validate norm
+        # First cast to int
+        norm = int(norm)
+        if norm < 1:
+            raise ValueError(
+                f"Invalid value for norm in constraints_by_jacobian_norm ({norm}). "
+                "Value must be a positive integer."
+            )
+
+        # We want to avoid generating the Jacobian and NLP more than once, so first we
+        # will identify the top-level block and collect all the constraints of interest
+        # as a list or iterator
+
+        if isinstance(blk_or_cons, BlockData):
+            # Scalar Block or element of Indexed Block
+            if descend_into:
+                # Scale all constraints, so pass con_list=None
+                con_list = None
+            else:
+                # Otherwise, get an iterator of constraints only on block
+                con_list = blk_or_cons.component_data_objects(
+                    Constraint, descend_into=False
+                )
+            jblock = blk_or_cons
+
+        elif isinstance(blk_or_cons, Block):
+            # Indexed Block
+            # Underlying tools do not work for Indexed blocks, so
+            # use parent block instead and collect constraints of interest
+            con_list = []
+            for b in blk_or_cons.values():
+                for c in b.component_data_objects(
+                    Constraint, descend_into=descend_into
+                ):
+                    con_list.append(c)
+            jblock = blk_or_cons.parent_block()
+
+        elif isinstance(blk_or_cons, ConstraintData):
+            # Scalar Constraint or element of Indexed Constraint
+            con_list = [blk_or_cons]
+            jblock = blk_or_cons.parent_block()
+
+        elif isinstance(blk_or_cons, Constraint):
+            # Indexed Constraint
+            con_list = blk_or_cons.values()
+            jblock = blk_or_cons.parent_block()
+
+        else:
+            raise TypeError(f"{blk_or_cons.name} is not a block or constraint.")
+
+        # Once we have a single target block and list of constraints, call the scaler method
+        # once for all the constraints
+        self._con_by_norm(jblock, con_list=con_list, norm=norm)
+
+    def _vardata_by_magnitude(self, vardata):
+        if value(vardata) is None:
+            sf = 1.0
```

Comment:
I don't think this is working as intended. I've tried replacing the scaling in the `demo_flowsheet.py` with 
```
    scaler = AutoScaler()
    scaler.variables_by_magnitude(m)
```
but I get the following error: `ValueError: No value for uninitialized NumericValue object fs.M01.minimum_pressure[0.0,1]`. I suspect this would occur anytime `value(vardata)` is None. I'm not sure if the solution is to not scale variables with a value of None or find a way to (temporarily?) fix its value

Reply:
Ah - the main thing is that this should be `vardata.value` and not `value(vardata)`.

Reply:
That fixes the issue